mesonbuild/dependencies/mpi.py

```diff
@@ -147,53 +161,34 @@ def _is_link_arg(self, f: str) -> bool:
                     f == '-pthread' or
                     (f.startswith('-W') and f != '-Wall' and not f.startswith('-Werror')))
 
-
-class IntelMPIConfigToolDependency(_MPIConfigToolDependency):
-
-    """Wrapper around Intel's mpiicc and friends."""
-
-    version_arg = '-v'  # --version is not the same as -v
-
-    def __init__(self, name: str, env: 'Environment', kwargs: T.Dict[str, T.Any],
-                 language: T.Optional[str] = None):
-        super().__init__(name, env, kwargs, language=language)
-        if not self.is_found:
-            return
-
-        args = self.get_config_value(['-show'], 'link and compile args')
-        self.compile_args = self._filter_compile_args(args)
-        self.link_args = self._filter_link_args(args)
-
-    def _sanitize_version(self, out: str) -> str:
-        v = re.search(r'(\d{4}) Update (\d)', out)
+    def _check_and_get_version(self, tool: T.List[str], returncode: int) -> T.Tuple[bool, T.Union[str, None]]:
+        p, out = Popen_safe(tool + ["--showme:version"])[:2]
+        valid = p.returncode == returncode
+        if valid:
+            # OpenMPI
+            v = re.search(r'\d+.\d+.\d+', out)
+            if v:
+                version = v.group(0)
+            else:
+                version = None
+            return valid, version
+
+        # --version is not the same as -v
+        p, out = Popen_safe(tool + ["-v"])[:2]
+        valid = p.returncode == returncode
+        out = out.split("\n", maxsplit=1)[0]
+
+        # cases like "mpicc for MPICH version 4.2.2"
+        v = re.search(r'\d+.\d+.\d+', out)
         if v:
-            return '{}.{}'.format(v.group(1), v.group(2))
-        return out
-
+            return valid, v.group(0)
 
-class OpenMPIConfigToolDependency(_MPIConfigToolDependency):
-
-    """Wrapper around OpenMPI mpicc and friends."""
-
-    version_arg = '--showme:version'
-
-    def __init__(self, name: str, env: 'Environment', kwargs: T.Dict[str, T.Any],
-                 language: T.Optional[str] = None):
-        super().__init__(name, env, kwargs, language=language)
-        if not self.is_found:
-            return
-
-        c_args = self.get_config_value(['--showme:compile'], 'compile_args')
-        self.compile_args = self._filter_compile_args(c_args)
-
-        l_args = self.get_config_value(['--showme:link'], 'link_args')
-        self.link_args = self._filter_link_args(l_args)
```

Comment:
We now resort to -show for openmpi as well, instead of the more granular openmpi interface?

Reply:
Indeed. I chose to get a first working implementation with only `-show`, which seems to be supported by all implementations (see https://github.com/mesonbuild/meson/issues/13615#issuecomment-2320541132).

But I'm going to add a commit using `--showme` for OpenMPI and `-compile_info` / `-link_info`. It's going to be cleaner (a bit more complicated though).

Reply:
Done in e744f1b4523ad309e6b04c69e30ff3c1da62dd27.